### PR TITLE
FOUNDATIONS: Modify Team

### LINKED ISSUE #97: FOUNDATIONS: Modify Team

### PR DESCRIPTION
**Pull request by jasurbekyusuf**
Closes #97

### CODE REVIEW

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs [hunk 1]
```diff
@@ -28,10 +26,50 @@ private void ValidateTeam(Team team)
...
                     secondDate: team.UpdatedDate,
                     secondDateName: nameof(Team.UpdatedDate)),
-
                 Parameter: nameof(Team.CreatedDate)));
         }
 
+        private void ValidateTeamOnModify(Team team)
+        {
+            ValidateTeamNotNull(team);
+            Validate(
```
Comment (jamshidbek001) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs:35]: need empty line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsNotSameAsCreatedDateAndLogItAsync()
+        {
+            // given
+            DateTimeOffset randomDateTime = GetRandomDateTime();
+            Team randomTeam = CreateRandomTeam(randomDateTime);
+            Team invalidTeam = randomTeam;
+            var invalidTeamException = new InvalidTeamException();
+
+            invalidTeamException.AddData(
+            key: nameof(Team.UpdatedDate),
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:132]: tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+                values: $"Date is the same as {nameof(Team.CreatedDate)}");
+
+            var expectedTeamValidationException =
+                new TeamValidationException(invalidTeamException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime())
+                    .Returns(randomDateTime);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:140]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        [Theory]
+        [MemberData(nameof(InvalidSeconds))]
+        public async Task ShouldThrowValidationExceptionOnModifyIfUpdatedDateIsNotRecentAndLogItAsync(int minutes)
+        {
+            // given
+            DateTimeOffset dateTime = GetRandomDateTime();
+            Team randomTeam = CreateRandomTeam(dateTime);
+            Team inputTeam = randomTeam;
+            inputTeam.UpdatedDate = dateTime.AddMinutes(minutes);
+            var invalidTeamException =
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:178]: 1 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+            invalidTeamException.AddData(
+                key: nameof(Team.UpdatedDate),
+                values: "Date is not recent.");
+
+            var expectedTeamValidatonException =
+                new TeamValidationException(invalidTeamException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime())
+                .Returns(dateTime);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:189]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+            TeamValidationException actualTeamValidationException =
+                await Assert.ThrowsAsync<TeamValidationException>(
+                    modifyTeamTask.AsTask);
+
+            // then
+            actualTeamValidationException.Should().BeEquivalentTo(expectedTeamValidatonException);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(),
+                    Times.Once);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:204]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedTeamValidatonException))),
+                        Times.Once);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:209]: 3 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTeamByIdAsync(It.IsAny<Guid>()),
+                    Times.Never);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:213]: 2 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+            invalidTeamException.AddData(
+                key: nameof(Team.CreatedDate),
+                values: $"Date is not same as {nameof(Team.CreatedDate)}.");
+
+            var expectedTeamValidationException =
+                new TeamValidationException(invalidTeamException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.SelectTeamByIdAsync(teamId))
+                .ReturnsAsync(storageTeam);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:292]: tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime())
+                .Returns(randomDateTimeOffset);
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:296]: 2 line or need tab

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs [hunk 11]
```diff
@@ -53,7 +53,6 @@ public static TheoryData<int> InvalidSeconds()
                 secondsInFuture
             };
         }
-
         private Expression<Func<Xeption, bool>> SameExceptionAs(Xeption expectedExceptoin) =>
```
Comment (jamshidbek001) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.cs:56]: need empty line

**Review by jamshidbek001 — APPROVED** [review]
please check comments

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs [hunk 12]
```diff
@@ -28,10 +26,51 @@ private void ValidateTeam(Team team)
...
+
+        private static void ValidateStorageTeam(Team maybeTeam, Guid teamId)
+        {
+            if (maybeTeam is null)
+            {
+                throw new NotFoundTeamException(teamId);
+            }
+        }
+
+        private static void ValidateAginstStorageTeamOnModify(Team inputTeam, Team storageTeam)
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs:58]: ValidateAginstStorageTeamOnModify =>
`ValidateAgainstStorageTeamOnModify` bo'lishi kk

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.cs [hunk 13]
```diff
@@ -34,5 +34,17 @@ public partial class TeamService : ITeamService
...
         });
+
+        public ValueTask<Team> ModifyTeamAsync(Team team) =>
+        TryCatch(async () =>
+        {
+            ValidateTeamOnModify(team);
+            var maybeTeam = await this.storageBroker.SelectTeamByIdAsync(team.Id);
+
+            ValidateStorageTeam(maybeTeam, team.Id);
+            ValidateAginstStorageTeamOnModify(inputTeam: team, storageTeam: maybeTeam);
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.cs:45]: Bu ikkalasi bir metod ichida bo'lishi kk

**Review by ElbekDeveloper — CHANGES_REQUESTED** [review]

**Review by jamshidbek001 — APPROVED** [review]

**Review by GulchexrasGithub — APPROVED** [review]
LGTM

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+                  modifyTeamTask.AsTask);
+
+            // then
+            actualTeamDependencyException.Should().BeEquivalentTo(
+                expectedTeamDependencyException);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
```
Comment (Javlonbek-dev) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:56]: storageBrokerMockni loggingBrokerMockdan keyin qo'ying

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.cs [hunk 15]
```diff
@@ -48,5 +48,15 @@ public partial class TeamService : ITeamService
 
             return maybeTeam;
         });
+
+        public ValueTask<Team> ModifyTeamAsync(Team team) =>
+        TryCatch(async () =>
+        {
+            ValidateTeamOnModify(team);
+            var maybeTeam = await this.storageBroker.SelectTeamByIdAsync(team.Id);
```
Comment (StudentofTATU) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.cs:56]: make 2 lines

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+                broker.GetCurrentDateTime())
+                    .Throws(sqlException);
+
+            // when
+            ValueTask<Team> modifyTeamTask =
+                this.teamService.ModifyTeamAsync(someTeam);
+
+            TeamDependencyException actualTeamDependencyException =
+              await Assert.ThrowsAsync<TeamDependencyException>(
+                  modifyTeamTask.AsTask);
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:46]: use TAB

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnModifyIfDatabaseUpdateExceptionOccursAndLogItAsync()
+        {
+            // given
+            int minutesInPast = GetRandomNegativeNumber();
+            DateTimeOffset randomDateTime = GetRandomDateTime();
+            Team randomTeam = CreateRandomTeam(randomDateTime);
+            Team SomeTeam = randomTeam;
+            Guid TeamId = SomeTeam.Id;
+            SomeTeam.CreatedDate = randomDateTime.AddMinutes(minutesInPast);
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:79]: change varialbe name to someTeam

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime())
+                    .Returns(randomDateTime);
+
+            // when
+            ValueTask<Team> modifyTeamTask =
+                this.teamService.ModifyTeamAsync(SomeTeam);
+
+            TeamDependencyException actualTeamDependencyException =
+              await Assert.ThrowsAsync<TeamDependencyException>(
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:101]: use TAB

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+                  modifyTeamTask.AsTask);
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:102]: use TAB

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+
+            // then
+            actualTeamDependencyException.Should().BeEquivalentTo(
+                expectedTeamDependencyException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTeamByIdAsync(TeamId), Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:111]: put this command  above of storageBrokerMock  at 108 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+                broker.GetCurrentDateTime(), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedTeamDependencyException))), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:119]: order

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+        [Fact]
+        public async Task ShouldThrowDependencyValidationExceptionOnModifyIfDatabaseUpdateConcurrencyErrorOccursAndLogItAsync()
+        {
+            // given
+            int minutesInPast = GetRandomNegativeNumber();
+            DateTimeOffset randomDateTime = GetRandomDateTime();
+            Team randomTeam = CreateRandomTeam(randomDateTime);
+            Team someTeam = randomTeam;
+            someTeam.CreatedDate = randomDateTime.AddMinutes(minutesInPast);
+            Guid TeamId = someTeam.Id;
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Exceptions.Modify.cs:131]: make variable name teamId

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.Modify.cs [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            // given
+            DateTimeOffset randomDate = GetRandomDateTime();
+            Team randomTeam = CreateRandomModifyTeam(randomDate);
+            Team inputTeam = randomTeam;
+            Team storageTeam = inputTeam.DeepClone();
+            storageTeam.UpdatedDate = randomTeam.CreatedDate;
+            Team updatedTeam = inputTeam;
+            Team expectedTeam = updatedTeam.DeepClone();
+            Guid TeamId = inputTeam.Id;
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.Modify.cs:29]: make variable name teamId

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 24]
```diff
@@ -0,0 +1,366 @@
...
+
+            //then
+            actualTeamValidationException.Should()
+                .BeEquivalentTo(expectedTeamValidationException);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedTeamValidationException))), Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:112]: this should be before loggingBrokerMock.Verify(..)  at 108 line

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedTeamValidationException))), Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTeamByIdAsync(teamId), Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
```
Comment (StudentofTATU) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:362]: order

**Review by StudentofTATU — CHANGES_REQUESTED** [review]
make some changes

**Review by StudentofTATU — APPROVED** [review]

**Comment by ElbekDeveloper** [comment]
@jasurbekyusuf conflicts

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 26]
```diff
@@ -56,6 +56,12 @@ private async ValueTask<Team> TryCatch(ReturningTeamFunction returningTeamFuncti
 
                 throw CreateAndDependencyValidationException(lockedTeamException);
             }
+            catch (DbUpdateException databaseUpdateException)
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs:59]: Buni nega `DbUpdateConcurrencyException` dan keyin qoyilganini bilasiz-a? Bilmasangiz, aytaman `DbUpdateException` dan inheritance oladi. Umumiy xatoliklar pastdan bo'lishi kerak.

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.Modify.cs [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        [Fact]
+        public async Task ShouldModifyTeamAsync()
+        {
+            // given
+            DateTimeOffset randomDate = GetRandomDateTime();
+            Team randomTeam = CreateRandomModifyTeam(randomDate);
+            Team inputTeam = randomTeam;
+            Team storageTeam = inputTeam.DeepClone();
+            storageTeam.UpdatedDate = randomTeam.CreatedDate;
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Logic.Modify.cs:26]: Bu shundog'am shu qiymatga egaku? Nima uchun bunday qilingan

File: Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs [hunk 28]
```diff
@@ -0,0 +1,367 @@
...
+            // given
+            int randomNumber = GetRandomNegativeNumber();
+            int randomMinutes = randomNumber;
+            DateTimeOffset randomDateTimeOffset = GetRandomDateTime();
+            Team randomTeam = CreateRandomModifyTeam(randomDateTimeOffset);
+            Team invalidTeam = randomTeam.DeepClone();
+            Team storageTeam = invalidTeam.DeepClone();
+            storageTeam.CreatedDate = storageTeam.CreatedDate.AddMinutes(randomMinutes);
+            storageTeam.UpdatedDate = storageTeam.UpdatedDate.AddMinutes(randomMinutes);
+            var invalidTeamException = new InvalidTeamException();
```
Comment (ElbekDeveloper) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/Teams/TeamServiceTests.Validations.Modify.cs:276]: Buni bir tekshirib ko'ring. Xatoga o'xshayapti...

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs [hunk 29]
```diff
@@ -84,12 +90,21 @@ private IQueryable<Team> TryCatch(ReturningTeamsFunction returningTeamsFunction)
...
+
+            return teamDependencyException;
+        }
+
+        private TeamServiceException CreateAndLogServiceException(
+            Exception exception)
         {
-            var teamServiceException = new TeamServiceException(exception);
-            this.loggingBroker.LogError(teamServiceException);
+            var TeamServiceException = new TeamServiceException(exception);
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs:104]: wrong naming

Comment (ElbekDeveloper) [reply Tarteeb.Api/Services/Foundations/Teams/TeamService.Exceptions.cs:104]: bu service exception dan pastda bo'ladi

File: Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs [hunk 30]
```diff
@@ -26,21 +26,56 @@ private void ValidateTeam(Team team)
...
 
-        private void ValidateStorageTeam(Team maybeTeam, Guid teamId)
+                (Rule: IsSame(
+                        firstDate: team.UpdatedDate,
+                        secondDate: team.CreatedDate,
+                        secondDateName: nameof(team.CreatedDate)),
+                     Parameter: nameof(team.UpdatedDate)));
+        }
+
+        private static void ValidateStorageTeam(Team maybeTeam, Guid teamId)
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Teams/TeamService.Validations.cs:50]: bu ValidateStorageTeamExists bo'lsin

**Review by ElbekDeveloper — CHANGES_REQUESTED** [review]

**Review by ElbekDeveloper — APPROVED** [review]

**Pull request merged**